From: George Wilkie <george.wilkie@intl.att.com>
Date: Wed, 1 Aug 2018 11:20:07 +0100
Subject: frr-reload-ldp

---
tools/frr-reload.py | 175 +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++------------------------------------
1 file changed, 121 insertions(+), 54 deletions(-)

diff --git a/tools/frr-reload.py b/tools/frr-reload.py
index 3e97635..b68613f 100755
--- a/tools/frr-reload.py
+++ b/tools/frr-reload.py
@@ -371,6 +371,21 @@ router ospf
  timers throttle spf 0 50 5000
 !
 end
+mpls ldp
+ !
+ address-family ipv4
+  discovery transport-address 1.1.1.1
+  !
+  interface dp0p1s2
+  end
+  !
+  interface dp0p1s3
+  end
+  !
+ exit-address-family
+ !
+!
+end
         '''
 
         # The code assumes that its working on the output from the "vtysh -m"
@@ -404,8 +419,8 @@ end
                                 "ip ",
                                 "ipv6 ",
                                 "log ",
-                                "mpls lsp",
-                                "mpls label",
+                                "mpls label ",
+                                "mpls lsp ",
                                 "no ",
                                 "password ",
                                 "ptm-enable",
@@ -438,13 +453,15 @@ end
                 ctx_keys = [line, ]
                 current_context_lines = []
 
-                log.debug('LINE %-50s: entering new context, %-50s', line, ctx_keys)
+                log.debug('LINE %-50s: entering new context, %-50s',
+                          line, ctx_keys)
                 self.save_contexts(ctx_keys, current_context_lines)
                 new_ctx = True
 
             elif line == "end":
                 self.save_contexts(ctx_keys, current_context_lines)
-                log.debug('LINE %-50s: exiting old context, %-50s', line, ctx_keys)
+                log.debug('LINE %-50s: exiting old context, %-50s',
+                          line, ctx_keys)
 
                 # Start a new context
                 new_ctx = True
@@ -455,7 +472,8 @@ end
             elif line == "exit-vrf":
                 self.save_contexts(ctx_keys, current_context_lines)
                 current_context_lines.append(line)
-                log.debug('LINE %-50s: append to current_context_lines, %-50s', line, ctx_keys)
+                log.debug(
+                    'LINE %-50s: append to current_context_lines, %-50s', line, ctx_keys)
 
                 #Start a new context
                 new_ctx = True
@@ -471,7 +489,8 @@ end
                     # Start a new context
                     ctx_keys = copy.deepcopy(main_ctx_key)
                     current_context_lines = []
-                    log.debug('LINE %-50s: popping from subcontext to ctx%-50s', line, ctx_keys)
+                    log.debug(
+                        'LINE %-50s: popping from subcontext to ctx%-50s', line, ctx_keys)
 
             elif line in ["exit-vni", "exit-ldp-if"]:
                 if sub_main_ctx_key:
@@ -480,7 +499,8 @@ end
                     # Start a new context
                     ctx_keys = copy.deepcopy(sub_main_ctx_key)
                     current_context_lines = []
-                    log.debug('LINE %-50s: popping from sub-subcontext to ctx%-50s', line, ctx_keys)
+                    log.debug(
+                        'LINE %-50s: popping from sub-subcontext to ctx%-50s', line, ctx_keys)
 
             elif new_ctx is True:
                 if not main_ctx_key:
@@ -491,7 +511,8 @@ end
 
                 current_context_lines = []
                 new_ctx = False
-                log.debug('LINE %-50s: entering new context, %-50s', line, ctx_keys)
+                log.debug('LINE %-50s: entering new context, %-50s',
+                          line, ctx_keys)
             elif (line.startswith("address-family ") or
                   line.startswith("vnc defaults") or
                   line.startswith("vnc l2-group") or
@@ -503,7 +524,8 @@ end
                 self.save_contexts(ctx_keys, current_context_lines)
                 current_context_lines = []
                 main_ctx_key = copy.deepcopy(ctx_keys)
-                log.debug('LINE %-50s: entering sub-context, append to ctx_keys', line)
+                log.debug(
+                    'LINE %-50s: entering sub-context, append to ctx_keys', line)
 
                 if line == "address-family ipv6" and not ctx_keys[0].startswith("mpls ldp"):
                     ctx_keys.append("address-family ipv6 unicast")
@@ -665,10 +687,12 @@ def get_normalized_ipv6_line(line):
                 try:
                     if 'ipaddress' not in sys.modules:
                         v6word = IPNetwork(word)
-                        norm_word = '%s/%s' % (v6word.network, v6word.prefixlen)
+                        norm_word = '%s/%s' % (v6word.network,
+                                                v6word.prefixlen)
                     else:
                         v6word = ip_network(word, strict=False)
-                        norm_word = '%s/%s' % (str(v6word.network_address), v6word.prefixlen)
+                        norm_word = '%s/%s' % (str(v6word.network_address),
+                                                v6word.prefixlen)
                 except ValueError:
                     pass
             if not norm_word:
@@ -742,28 +766,34 @@ def ignore_delete_re_add_lines(lines_to_add, lines_to_del):
                         peergroup = re_swpx_int_v6only_peergroup.group(2)
                         swpx_interface = "neighbor %s interface v6only" % swpx
 
-                    swpx_peergroup = "neighbor %s peer-group %s" % (swpx, peergroup)
-                    found_add_swpx_interface = line_exist(lines_to_add, ctx_keys, swpx_interface)
-                    found_add_swpx_peergroup = line_exist(lines_to_add, ctx_keys, swpx_peergroup)
+                    swpx_peergroup = "neighbor %s peer-group %s" % (
+						swpx, peergroup)
+                    found_add_swpx_interface = line_exist(
+						lines_to_add, ctx_keys, swpx_interface)
+                    found_add_swpx_peergroup = line_exist(
+						lines_to_add, ctx_keys, swpx_peergroup)
                     tmp_ctx_keys = tuple(list(ctx_keys))
 
                     if not found_add_swpx_peergroup:
                         tmp_ctx_keys = list(ctx_keys)
                         tmp_ctx_keys.append('address-family ipv4 unicast')
                         tmp_ctx_keys = tuple(tmp_ctx_keys)
-                        found_add_swpx_peergroup = line_exist(lines_to_add, tmp_ctx_keys, swpx_peergroup)
+                        found_add_swpx_peergroup = line_exist(
+							lines_to_add, tmp_ctx_keys, swpx_peergroup)
 
                         if not found_add_swpx_peergroup:
                             tmp_ctx_keys = list(ctx_keys)
                             tmp_ctx_keys.append('address-family ipv6 unicast')
                             tmp_ctx_keys = tuple(tmp_ctx_keys)
-                            found_add_swpx_peergroup = line_exist(lines_to_add, tmp_ctx_keys, swpx_peergroup)
+                            found_add_swpx_peergroup = line_exist(
+								lines_to_add, tmp_ctx_keys, swpx_peergroup)
 
                     if found_add_swpx_interface and found_add_swpx_peergroup:
                         deleted = True
                         lines_to_del_to_del.append((ctx_keys, line))
                         lines_to_add_to_del.append((ctx_keys, swpx_interface))
-                        lines_to_add_to_del.append((tmp_ctx_keys, swpx_peergroup))
+                        lines_to_add_to_del.append(
+							(tmp_ctx_keys, swpx_peergroup))
 
                 '''
                 Changing the bfd timers on neighbors is allowed without doing
@@ -771,7 +801,8 @@ def ignore_delete_re_add_lines(lines_to_add, lines_to_del):
                 will cause the peer to bounce unnecessarily, just skip the delete
                 and just do the add.
                 '''
-                re_nbr_bfd_timers = re.search(r'neighbor (\S+) bfd (\S+) (\S+) (\S+)', line)
+                re_nbr_bfd_timers = re.search(
+					r'neighbor (\S+) bfd (\S+) (\S+) (\S+)', line)
 
                 if re_nbr_bfd_timers:
                     nbr = re_nbr_bfd_timers.group(1)
@@ -810,8 +841,10 @@ def ignore_delete_re_add_lines(lines_to_add, lines_to_del):
 
                 If so then chop the del line and the corresponding add lines
                 '''
-                re_swpx_int_remoteas = re.search('neighbor (\S+) interface remote-as (\S+)', line)
-                re_swpx_int_v6only_remoteas = re.search('neighbor (\S+) interface v6only remote-as (\S+)', line)
+                re_swpx_int_remoteas = re.search(
+					'neighbor (\S+) interface remote-as (\S+)', line)
+                re_swpx_int_v6only_remoteas = re.search(
+				'neighbor (\S+) interface v6only remote-as (\S+)', line)
 
                 if re_swpx_int_remoteas or re_swpx_int_v6only_remoteas:
                     swpx_interface = None
@@ -826,16 +859,20 @@ def ignore_delete_re_add_lines(lines_to_add, lines_to_del):
                         remoteas = re_swpx_int_v6only_remoteas.group(2)
                         swpx_interface = "neighbor %s interface v6only" % swpx
 
-                    swpx_remoteas = "neighbor %s remote-as %s" % (swpx, remoteas)
-                    found_add_swpx_interface = line_exist(lines_to_add, ctx_keys, swpx_interface)
-                    found_add_swpx_remoteas = line_exist(lines_to_add, ctx_keys, swpx_remoteas)
+                    swpx_remoteas = "neighbor %s remote-as %s" % (
+						swpx, remoteas)
+                    found_add_swpx_interface = line_exist(
+						lines_to_add, ctx_keys, swpx_interface)
+                    found_add_swpx_remoteas = line_exist(
+						lines_to_add, ctx_keys, swpx_remoteas)
                     tmp_ctx_keys = tuple(list(ctx_keys))
 
                     if found_add_swpx_interface and found_add_swpx_remoteas:
                         deleted = True
                         lines_to_del_to_del.append((ctx_keys, line))
                         lines_to_add_to_del.append((ctx_keys, swpx_interface))
-                        lines_to_add_to_del.append((tmp_ctx_keys, swpx_remoteas))
+                        lines_to_add_to_del.append(
+							(tmp_ctx_keys, swpx_remoteas))
 
             '''
             We made the 'bgp bestpath as-path multipath-relax' command
@@ -847,9 +884,11 @@ def ignore_delete_re_add_lines(lines_to_add, lines_to_del):
             resets.
             '''
             if 'multipath-relax' in line:
-                re_asrelax_new = re.search('^bgp\s+bestpath\s+as-path\s+multipath-relax$', line)
+                re_asrelax_new = re.search(
+					'^bgp\s+bestpath\s+as-path\s+multipath-relax$', line)
                 old_asrelax_cmd = 'bgp bestpath as-path multipath-relax no-as-set'
-                found_asrelax_old = line_exist(lines_to_add, ctx_keys, old_asrelax_cmd)
+                found_asrelax_old = line_exist(
+					lines_to_add, ctx_keys, old_asrelax_cmd)
 
                 if re_asrelax_new and found_asrelax_old:
                     deleted = True
@@ -863,7 +902,8 @@ def ignore_delete_re_add_lines(lines_to_add, lines_to_del):
             is issued.
             '''
             if line.startswith('table-map'):
-                found_table_map = line_exist(lines_to_add, ctx_keys, 'table-map', False)
+                found_table_map = line_exist(
+					lines_to_add, ctx_keys, 'table-map', False)
 
                 if found_table_map:
                     lines_to_del_to_del.append((ctx_keys, line))
@@ -880,7 +920,8 @@ def ignore_delete_re_add_lines(lines_to_add, lines_to_del):
             table_num = re_importtbl.group(1)
             for ctx in lines_to_add:
                 if ctx[0][0].startswith('ip import-table %s distance' % table_num):
-                    lines_to_del_to_del.append((('ip import-table %s' % table_num,), None))
+                    lines_to_del_to_del.append(
+						(('ip import-table %s' % table_num,), None))
                     lines_to_add_to_del.append((ctx[0], None))
 
         '''
@@ -908,7 +949,8 @@ def ignore_delete_re_add_lines(lines_to_add, lines_to_del):
             ctx_keys[1] == 'address-family l2vpn evpn' and
             ctx_keys[2].startswith('vni')):
 
-            re_route_target = re.search('^route-target import (.*)$', line) if line is not None else False
+            re_route_target = re.search(
+				'^route-target import (.*)$', line) if line is not None else False
 
             if re_route_target:
                 rt = re_route_target.group(1).strip()
@@ -916,8 +958,10 @@ def ignore_delete_re_add_lines(lines_to_add, lines_to_del):
                 route_target_export_line = "route-target export %s" % rt
                 route_target_both_line = "route-target both %s" % rt
 
-                found_route_target_export_line = line_exist(lines_to_del, ctx_keys, route_target_export_line)
-                found_route_target_both_line = line_exist(lines_to_add, ctx_keys, route_target_both_line)
+                found_route_target_export_line = line_exist(
+					lines_to_del, ctx_keys, route_target_export_line)
+                found_route_target_both_line = line_exist(
+					lines_to_add, ctx_keys, route_target_both_line)
 
                 '''
                 If the running configs has
@@ -930,9 +974,12 @@ def ignore_delete_re_add_lines(lines_to_add, lines_to_del):
                 then we can ignore deleting the import/export and ignore adding the 'both'
                 '''
                 if found_route_target_export_line and found_route_target_both_line:
-                    lines_to_del_to_del.append((ctx_keys, route_target_import_line))
-                    lines_to_del_to_del.append((ctx_keys, route_target_export_line))
-                    lines_to_add_to_del.append((ctx_keys, route_target_both_line))
+                    lines_to_del_to_del.append(
+						(ctx_keys, route_target_import_line))
+                    lines_to_del_to_del.append(
+						(ctx_keys, route_target_export_line))
+                    lines_to_add_to_del.append(
+						(ctx_keys, route_target_both_line))
 
         if not deleted:
             found_add_line = line_exist(lines_to_add, ctx_keys, line)
@@ -962,7 +1009,8 @@ def ignore_delete_re_add_lines(lines_to_add, lines_to_del):
                     tmp_ctx_keys = list(ctx_keys)[:-1]
                     tmp_ctx_keys = tuple(tmp_ctx_keys)
 
-                    found_add_line = line_exist(lines_to_add, tmp_ctx_keys, line)
+                    found_add_line = line_exist(
+						lines_to_add, tmp_ctx_keys, line)
 
                     if found_add_line:
                         lines_to_del_to_del.append((ctx_keys, line))
@@ -1089,8 +1137,10 @@ def compare_context_objects(newconf, running):
             for line in newconf_ctx.lines:
                 lines_to_add.append((newconf_ctx_keys, line))
 
-    (lines_to_add, lines_to_del) = ignore_delete_re_add_lines(lines_to_add, lines_to_del)
-    (lines_to_add, lines_to_del) = ignore_unconfigurable_lines(lines_to_add, lines_to_del)
+    (lines_to_add, lines_to_del) = ignore_delete_re_add_lines(
+		lines_to_add, lines_to_del)
+    (lines_to_add, lines_to_del) = ignore_unconfigurable_lines(
+		lines_to_add, lines_to_del)
 
     return (lines_to_add, lines_to_del)
 
@@ -1123,19 +1173,30 @@ def vtysh_config_available(bindir, confdir):
 
 if __name__ == '__main__':
     # Command line options
-    parser = argparse.ArgumentParser(description='Dynamically apply diff in frr configs')
-    parser.add_argument('--input', help='Read running config from file instead of "show running"')
+    parser = argparse.ArgumentParser(
+		description='Dynamically apply diff in frr configs')
+    parser.add_argument(
+		'--input', help='Read running config from file instead of "show running"')
     group = parser.add_mutually_exclusive_group(required=True)
-    group.add_argument('--reload', action='store_true', help='Apply the deltas', default=False)
-    group.add_argument('--test', action='store_true', help='Show the deltas', default=False)
-    parser.add_argument('--debug', action='store_true', help='Enable debugs', default=False)
-    parser.add_argument('--stdout', action='store_true', help='Log to STDOUT', default=False)
+    group.add_argument('--reload', action='store_true',
+						help='Apply the deltas', default=False)
+    group.add_argument('--test', action='store_true',
+						help='Show the deltas', default=False)
+    parser.add_argument('--debug', action='store_true',
+						help='Enable debugs', default=False)
+    parser.add_argument('--stdout', action='store_true',
+						help='Log to STDOUT', default=False)
     parser.add_argument('filename', help='Location of new frr config file')
-    parser.add_argument('--overwrite', action='store_true', help='Overwrite frr.conf with running config output', default=False)
-    parser.add_argument('--bindir', help='path to the vtysh executable', default='/usr/bin')
-    parser.add_argument('--confdir', help='path to the daemon config files', default='/etc/frr')
-    parser.add_argument('--rundir', help='path for the temp config file', default='/var/run/frr')
-    parser.add_argument('--daemon', help='daemon for which want to replace the config', default='')
+    parser.add_argument('--overwrite', action='store_true',
+                help='Overwrite frr.conf with running config output', default=False)
+    parser.add_argument('--bindir',
+                help='path to the vtysh executable', default='/usr/bin')
+    parser.add_argument('--confdir',
+                help='path to the daemon config files', default='/etc/frr')
+    parser.add_argument('--rundir',
+                help='path for the temp config file', default='/var/run/frr')
+    parser.add_argument('--daemon',
+                help='daemon for which want to replace the config', default='')
 
     args = parser.parse_args()
 
@@ -1147,8 +1208,10 @@ if __name__ == '__main__':
                             format='%(asctime)s %(levelname)5s: %(message)s')
 
         # Color the errors and warnings in red
-        logging.addLevelName(logging.ERROR, "\033[91m  %s\033[0m" % logging.getLevelName(logging.ERROR))
-        logging.addLevelName(logging.WARNING, "\033[91m%s\033[0m" % logging.getLevelName(logging.WARNING))
+        logging.addLevelName(
+			logging.ERROR, "\033[91m  %s\033[0m" % logging.getLevelName(logging.ERROR))
+        logging.addLevelName(
+			logging.WARNING, "\033[91m%s\033[0m" % logging.getLevelName(logging.WARNING))
 
     elif args.reload:
         if not os.path.isdir('/var/log/frr/'):
@@ -1313,7 +1376,8 @@ if __name__ == '__main__':
         for x in range(2):
             running = Config()
             running.load_from_show_running(args.bindir, args.confdir, args.daemon)
-            log.debug('Running Frr Config (Pass #%d)\n%s', x, running.get_lines())
+            log.debug('Running Frr Config (Pass #%d)\n%s',
+						x, running.get_lines())
 
             (lines_to_add, lines_to_del) = compare_context_objects(newconf, running)
 
@@ -1375,7 +1439,8 @@ if __name__ == '__main__':
                             last_arg = cmd[-1].split(' ')
 
                             if len(last_arg) <= 2:
-                                log.error('"%s" we failed to remove this command', original_cmd)
+                                log.error(
+                                '"%s" we failed to remove this command', original_cmd)
                                 break
 
                             new_last_arg = last_arg[0:-1]
@@ -1401,14 +1466,16 @@ if __name__ == '__main__':
                                             string.digits) for _ in range(6))
 
                     filename = args.rundir + "/reload-%s.txt" % random_string
-                    log.info("%s content\n%s" % (filename, pformat(lines_to_configure)))
+                    log.info("%s content\n%s" % (
+                            filename, pformat(lines_to_configure)))
 
                     with open(filename, 'w') as fh:
                         for line in lines_to_configure:
                             fh.write(line + '\n')
 
                     try:
-                        subprocess.check_output([str(args.bindir + '/vtysh'), '--config_dir', args.confdir, '-f', filename])
+                        subprocess.check_output(
+                            [str(args.bindir + '/vtysh'), '--config_dir', args.confdir, '-f', filename])
                     except subprocess.CalledProcessError as e:
                         log.warning("frr-reload.py failed due to\n%s" % e.output)
                         reload_ok = False
