--- frr.orig/tools/frr.service
+++ frr/tools/frr.service
@@ -1,9 +1,8 @@
 [Unit]
 Description=FRRouting
 Documentation=https://frrouting.readthedocs.io/en/latest/setup.html
-Wants=network.target
-After=network-pre.target systemd-sysctl.service
-Before=network.target
+PartOf=vyatta-routing-frr.target
+Before=vyatta-routing-frr-early.target
 OnFailure=heartbeat-failed@%n.service
 
 [Service]
--- frr.orig/tools/etc/frr/daemons
+++ frr/tools/etc/frr/daemons
@@ -14,14 +14,14 @@
 #
 # The watchfrr and zebra daemons are always started.
 #
-bgpd=no
-ospfd=no
-ospf6d=no
+bgpd=yes
+ospfd=yes
+ospf6d=yes
 ripd=no
 ripngd=no
 isisd=no
 pimd=no
-ldpd=no
+ldpd=yes
 nhrpd=no
 eigrpd=no
 babeld=no
@@ -37,7 +37,7 @@ vrrpd=no
 # Check /etc/pam.d/frr if you intend to use "vtysh"!
 #
 vtysh_enable=yes
-zebra_options="  -A 127.0.0.1 -s 90000000"
+zebra_options="  -A 127.0.0.1 -s 90000000 -M fpm"
 bgpd_options="   -A 127.0.0.1"
 ospfd_options="  -A 127.0.0.1"
 ospf6d_options=" -A ::1"
