From: George Wilkie <george.wilkie@intl.att.com>
Date: Wed, 1 Aug 2018 11:20:07 +0100
Subject: syslog_debug

---
 lib/zlog_targets.c |    2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

--- a/lib/zlog_targets.c
+++ b/lib/zlog_targets.c
@@ -500,7 +500,7 @@ static void zlog_syslog_sigsafe(struct z
 static pthread_mutex_t syslog_cfg_mutex = PTHREAD_MUTEX_INITIALIZER;
 static struct zlt_syslog *zlt_syslog;
 static int syslog_facility = LOG_DAEMON;
-static int syslog_prio_min = ZLOG_DISABLED;
+static int syslog_prio_min = LOG_DEBUG;
 
 void zlog_syslog_set_facility(int facility)
 {
